### PR TITLE
show an error if gopmctl start or restart fails

### PR DESCRIPTION
**Pull request by rogpeppe**
This also gives us an opportunity to experiment with returning
rich errors from gRPC calls, which paves the way for returning
syntax errors that can be formatted nicely on the client side.

Also fix an error in the stop logic that would leave a process in Backoff
state when it should be moved to Stopped.

Also implement the `oneshot` attribute and improve RPC docs.
See the commit for details.

### CODE REVIEW

File: rpc/error.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	return e.Code
+}
+
+// ServerErrors returns a server option that converts protobuf-marshalable errors
+// into errors that can be passed back over the wire and reassembled with ClientErrors.
+// Pass this to grpc.NewServer before registering the gRPC server implementation.
+//
+// Any error returned from an RPC endpoint that implements proto.Message
+// will be added as error details to the gRPC error.
+func ServerErrors() grpc.ServerOption {
```
Comment (stuartcarnie) [inline rpc/error.go:103]: ✅

**Review by stuartcarnie — COMMENTED** [review]
I like that your solution transparently handles errors via an interceptor for the server and client. I'd like for us to document this expectation, but I'm not sure where it should go.

### An aside

Passing rich error messages over gRPC is something that has been a thorn in the side of many of us for a long time, so I took some time to read up on it again via https://grpc.io/docs/guides/error/. In the section titled [Richer error model](https://grpc.io/docs/guides/error/#richer-error-model), it provides [a link](https://cloud.google.com/apis/design/errors#error_model) to a richer error model utilised by Google. In particular, they define a [specific set of protobuf messages](https://cloud.google.com/apis/design/errors#error_details) that a client can expect in the `Details` of the grpc.Status message:

> Here are some example error_details payloads:
>
> * `ErrorInfo`: Provides structured error information that is both stable and extensible.
> * `RetryInfo`: Describes when clients can retry a failed request, may be returned on `Code.UNAVAILABLE` or `Code.ABORTED`
> `QuotaFailure`: Describes how a quota check failed, may be returned on `Code.RESOURCE_EXHAUSTED`
> `BadRequest`: Describes violations in a client request, may be returned on `Code.INVALID_ARGUMENT`

My primary point for sharing these links is that the prior art is helpful for forming our own ideas and guidance for _developing a specification_. We don't necessarily have to use Google's, but we should formalise how errors are propagated and how a client should be expected to handle them. We can use this as an experiment for some of our own services at InfluxData.

The [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto) file defines the complete list of messages a client could expect to find in the Details slice. `ErrorInfo` and `LocalizedMessage` seem potentially useful.

----

Incidentally, I've been experimenting with a Swift gRPC client and macOS SwiftUI client.

**Comment by rogpeppe** [comment]
@stuartcarnie 

> We don't necessarily have to use Google's, but we should formalise how errors are propagated and how a client should be expected to handle them.

My thought is to document possible error details on a per-call basis, much like we do (or _should_ do!) in Go.

Alternatively (or, better, additionally) we can document which proto messages function as error returns, so a non-Go client that doesn't have the benefit of the annotation provided by the `Error` method can do a switch on the details contents, or whatever might be appropriate.

I'll make those changes now to illustrate.

**Comment by rogpeppe** [comment]
I've just added a commit that adds `oneshot` support, so that it can avoid conflicts, because it depends on this PR.

File: config/schema.cue [hunk 2]
```diff
@@ -76,6 +76,12 @@ config:  #Config
 	// a pseudo-target for a bunch of other dependencies.
 	command: string
 
+	// oneshot specifies that the command is always intended
+	// to run to completion. If this is true, start_seconds is ignored and
+	// the program will never enter the Running state - programs that
+	// depend on it will be started only when the command exits successfully.
+	oneshot?: bool
+
```
Comment (stuartcarnie) [inline config/schema.cue:84]: 🥳

**Review by stuartcarnie — APPROVED** [review]
This looks good – a `oneshot` process will be very handy.

What happens when a `oneshot` process is started again? Will it restart all the dependent processes? I suspect that the updated dependency PR will be responsible for addressing this?

**Comment by rogpeppe** [comment]
> What happens when a oneshot process is started again? Will it restart all the dependent processes?

That's an excellent question. Currently it won't restart processes that depend on it, although it will start them after the oneshot process has completed if they're not already started. Maybe we should consider that a start of an exited process to be the same as a restart, and therefore tear down all processes that depend on it before starting again. That's something for another PR though - I think it should be straightforward to do.

**Pull request merged**